port/cpl_vsil_curl.cpp

```diff
@@ -4880,6 +4893,84 @@ int VSICurlParseUnixPermissions(const char* pszPermissions)
     return nMode;
 }
 
+/************************************************************************/
+/*                  Cache of file properties.                           */
+/************************************************************************/
+
+static std::mutex oCacheFilePropMutex;
+static lru11::Cache<std::string, FileProp>* poCacheFileProp = nullptr;
+
+/************************************************************************/
+/*                   VSICURLGetCachedFileProp()                         */
+/************************************************************************/
+
+bool VSICURLGetCachedFileProp( const char* pszURL, FileProp& oFileProp )
+{
+    std::lock_guard<std::mutex> oLock(oCacheFilePropMutex);
+    return poCacheFileProp != nullptr &&
+           poCacheFileProp->tryGet(std::string(pszURL), oFileProp) &&
+            // Let a chance to use new auth parameters
+           !(oFileProp.eExists == EXIST_NO &&
+             gnGenerationAuthParameters != oFileProp.nGenerationAuthParameters);
+}
+
+/************************************************************************/
+/*                   VSICURLSetCachedFileProp()                         */
+/************************************************************************/
+
+void VSICURLSetCachedFileProp( const char* pszURL, FileProp& oFileProp )
+{
+    std::lock_guard<std::mutex> oLock(oCacheFilePropMutex);
+    if( poCacheFileProp == nullptr )
+        poCacheFileProp = new lru11::Cache<std::string, FileProp>(100 * 1024);
+    oFileProp.nGenerationAuthParameters = gnGenerationAuthParameters;
+    poCacheFileProp->insert(std::string(pszURL), oFileProp);
+}
+
+/************************************************************************/
+/*                   VSICURLInvalidateCachedFileProp()                  */
+/************************************************************************/
+
+void VSICURLInvalidateCachedFileProp( const char* pszURL )
+{
+    std::lock_guard<std::mutex> oLock(oCacheFilePropMutex);
+    if( poCacheFileProp != nullptr )
+        poCacheFileProp->remove(std::string(pszURL));
+}
+
+/************************************************************************/
+/*               VSICURLInvalidateCachedFilePropPrefix()                */
+/************************************************************************/
+
+void VSICURLInvalidateCachedFilePropPrefix( const char* pszURL )
+{
+    std::lock_guard<std::mutex> oLock(oCacheFilePropMutex);
+    if( poCacheFileProp != nullptr )
+    {
+        std::list<std::string> keysToRemove;
+        const size_t nURLSize = strlen(pszURL);
+        auto lambda = [&keysToRemove, &pszURL, nURLSize](
+            const lru11::KeyValuePair<std::string, FileProp>& kv)
+        {
+            if( strncmp(kv.key.c_str(), pszURL, nURLSize) == 0 )
+                keysToRemove.push_back(kv.key);
+        };
+        poCacheFileProp->cwalk(lambda);
+        for( auto& key: keysToRemove )
```

Comment:
is the intention of separating the visiting from the removal here to avoid modifying the props while they are being visited?

Reply:
yes, I'm not completely sure that removing while visiting is safe here.